refactor!: Rename Logger to LogSink

arv opened the pull request:
And make LogContext take a LogSink instead of an OptionalLogger

arv committed:
refactor!: Rename Logger to LogSink

And make LogContext take a LogSink instead of an OptionalLogger

arv commented on src/logger.ts:
```diff
@@ -90,48 +84,59 @@ export class SilentLogger implements OptionalLogger {
-    );
-    this._s = tag;
-    this._logger = actualLogger;
+  constructor();
+  constructor(level: LogLevel, tag?: string);
+  constructor(logSink: LogSink, level?: LogLevel, tag?: string);
+  constructor(
+    levelOrLogSink?: LogLevel | LogSink,
+    levelOrTag?: LogLevel | string,
+    tag?: string,
+  ) {
+    let logSink: LogSink;
+    let level: LogLevel;
+
+    if (!levelOrLogSink) {
+      // No args: constructor();
+      logSink = consoleLogSink;
+      level = 'info';
+    } else if (isLogLevel(levelOrLogSink)) {
+      // No sink: constructor(level: LogLevel, tag?: string);
+      logSink = consoleLogSink;
+      level = levelOrLogSink;
+      tag = levelOrTag;
+    } else {
+      // constructor(logSink: LogSink, level?: LogLevel, tag?: string);
+      logSink = levelOrLogSink;
+      if (levelOrTag) {
+        level = levelOrTag as LogLevel;
+      } else {
+        level = 'info';
+      }
+      tag ??= '';
+    }
```
This is a lot of goop to support all the overrides. Maybe we should get rid of some of them?

phritz replied:
It is a lot of goop. Can we not require log level and logsink and an optional tag? I find the defaults and overrides to be too much magic, would rather write a tiny wrapper to invoke the constructor with reasonable defaults than to have this stuff in the constructor itself.

aboodman replied:
You could also make them mutable fields so they can be set after construction.

How about:

```ts
constructor(level: string)
tag?: string|undefined;
sinks: LogSink[]; // default: just ConsoleSink
```

Question: why do we have a tag in addition to context? Why not have the tag be the first context?

PS bike shed should be pink.

arv replied:
Yeah, tag is just the first context. Can be removed.

Making them public properties is a bit more complicated because then they can change

I'm going to trim down the crazies.

arv commented on src/logger.ts:
```diff
@@ -141,3 +146,21 @@ function isLogLevel(v: unknown): v is LogLevel {
+
+  if (logSink.flush) {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    newLogSink.flush = () => logSink.flush!();
```
I realize that there is no test for flush... Adding

arv reviewed (commented):
I hope this makes a little more sense now

arv commented:
Actually, the flush does not need to be forwarded... Undoing that.

arv committed:
No need to forward flsuh

phritz reviewed (approved):
I like the name change and think you should just merge this because it is a helpful change for a few things I would like to do.

arv committed:
Make the arguments to LogContext less crazy

Made them all optional but changed the order a bit to make it easier to use:

```ts
constructor(
  level: LogLevel = 'info',
  logSink: LogSink = consoleLogSink,
  tag = '',
) {
```

We still need the tag to chain things together, even thought that could
have been done in other ways.

arv committed:
Simplify further

arv commented:
Tagged and pushed to npm
